types/jquery/misc.d.ts

```diff
@@ -4704,17 +4631,173 @@ $( "ul" ).click( handler ).find( "ul" ).hide();
+```html
+<!doctype html>
+<html lang="en">
+<head>
+  <meta charset="utf-8">
+  <title>event.data demo</title>
+  <script src="https://code.jquery.com/jquery-3.3.1.js"></script>
+</head>
+<body>
+​
+<button> 0 </button>
+<button> 1 </button>
+<button> 2 </button>
+<button> 3 </button>
+<button> 4 </button>
+​
+<div id="log"></div>
+​
+<script>
+var logDiv = $( "#log" );
+​
+for ( var i = 0; i < 5; i++ ) {
+  $( "button" ).eq( i ).on( "click", { value: i }, function( event ) {
+    var msgs = [
+      "button = " + $( this ).index(),
+      "event.data.value = " + event.data.value,
+      "i = " + i
+    ];
+    logDiv.append( msgs.join( ", " ) + "<br>" );
+  });
+}
+</script>
+​
+</body>
+</html>
+```
+        */
+        data: TData;
 
-    interface EventHandler<TCurrentTarget, TData = null> extends EventHandlerBase<TCurrentTarget, Event<TCurrentTarget, TData>> { }
+        /**
+         * The namespace specified when the event was triggered.
+         * @see \`{@link https://api.jquery.com/event.namespace/ }\`
+         * @since 1.4.3
+         * @example ​ ````Determine the event namespace used.
+```html
+<!doctype html>
+<html lang="en">
+<head>
+  <meta charset="utf-8">
+  <title>event.namespace demo</title>
+  <script src="https://code.jquery.com/jquery-3.3.1.js"></script>
+</head>
+<body>
+​
+<button>display event.namespace</button>
+<p></p>
+​
+<script>
+$( "p" ).on( "test.something", function( event ) {
+  alert( event.namespace );
+});
+$( "button" ).click(function( event ) {
+  $( "p" ).trigger( "test.something" );
+});
+</script>
+​
+</body>
+</html>
+```
+         */
+        namespace?: string;
+        /**
+         * The last value returned by an event handler that was triggered by this event, unless the value was undefined.
+         * @see \`{@link https://api.jquery.com/event.result/ }\`
+         * @since 1.3
+         * @example ​ ````Display previous handler&#39;s return value
+```html
+<!doctype html>
+<html lang="en">
+<head>
+  <meta charset="utf-8">
+  <title>event.result demo</title>
+  <script src="https://code.jquery.com/jquery-3.3.1.js"></script>
+</head>
+<body>
+​
+<button>display event.result</button>
+<p></p>
+​
+<script>
+$( "button" ).click(function( event ) {
+  return "hey";
+});
+$( "button" ).click(function( event ) {
+  $( "p" ).html( event.result );
+});
+</script>
+​
+</body>
+</html>
+```
+         */
+        result?: any;
+    }
+
+    interface TypeToTriggeredEventMap<
+        TDelegateTarget,
+        TData,
+        TCurrentTarget,
+        TTarget
+    > {
+        [type: string]: TriggeredEvent<TDelegateTarget, TData, TCurrentTarget, TTarget>;
+    }
 
     // Extra parameters can be passed from trigger()
     type EventHandlerBase<TContext, T> = (this: TContext, t: T, ...args: any[]) => any;
 
+    type EventHandler<
+        TCurrentTarget,
+        TData = undefined
+    > = EventHandlerBase<TCurrentTarget, TriggeredEvent<TCurrentTarget, TData>>;
+
+    type TypeEventHandler<
```

Comment:
I can't find any instantiations of `TypeEventHandler` where `TCurrentTarget`, `TTarget` and `TContext` aren't all exactly the same type. It would be simpler to just have one type argument (`TTarget`?) instead of 3 separate ones.

Reply:
This comment was for an earlier commit -- I see you removed `TContext`, but my comment still applies to `TCurrentTarget` and `TTarget`. They could be simplified.

Reply:
Sounds like a possible mistake elsewhere, an event's currentTarget is definitely not always the same as its target.

The way React dealt with it is that Change events and others that don't bubble have target = currentTarget, but other events have target being EventTarget.

TContext is supposed to be whatever was the 2nd argument of .on in 3-argument form so sounds like another omission.

Reply:
For directly bound event handlers, `TElement` is passed for `TCurrentTarget` and `TTarget` because they are known to be the same object. For delegated event handlers, `any` is passed for `TCurrentTarget` and `TTarget` because they can be any element in the event propagation path and potentially refer to different objects.

I went with 2 type parameters because they're not semantically the same. I was also concerned there might be scenarios where say a user asserts that `currentTarget` is of a certain type which might allow TypeScript to infer that `target` is the same type since they share a type parameter. I don't know if TypeScript actually does this but it didn't seem unreasonable. There's also concern with what to name it. `jQuery.Event` has 4 properties that have `target` in the name, one of which is named `target`. I think if a user came across a type parameter named `TTarget`, they'd either expect it to cover just the `target` property or all 4 `*Target` properties, not just a particular 2 of the 4.